python/src/spark_rapids_ml/knn.py

```diff
@@ -590,32 +701,534 @@ async def do_allGather() -> List[str]:
+
+        knn_df: pyspark.sql.DataFrame
+            the result k approximate nearest neighbors (ANNs) dataframe that has three
+            columns (id, indices, distances). Each row of knn_df corresponds to the k-ANNs
+            result of a query vector, identified by the id column. The indices/distances
+            column stores the ids/distances of knn item_df vectors.
+        """
+
+        query_df_withid = self._ensureIdCol(query_df)
+        self.bcast_qids, self.bcast_qfeatures = self._broadcast_as_nparray(
+            query_df_withid
+        )
+
+        knn_df = self._transform_evaluate_internal(
+            self._item_df_withid, schema=self._out_schema()
+        )
+        k = self.getK()
+
+        query_id_col_name = f"query_{self.getIdCol()}"
+        knn_df_agg = self.__class__._agg_topk(
+            knn_df, query_id_col_name, "indices", "distances", k
+        )
+
+        return (self._item_df_withid, query_df_withid, knn_df_agg)
+
     def _get_cuml_transform_func(
         self, dataset: DataFrame, eval_metric_info: Optional[EvalMetricInfo] = None
     ) -> Tuple[
         _ConstructFunc,
         _TransformFunc,
         Optional[_EvaluateFunc],
     ]:
-        raise NotImplementedError(
-            "'_CumlModel._get_cuml_transform_func' method is not implemented. Use 'kneighbors' instead."
-        )
 
-    def exactNearestNeighborsJoin(
+        cuml_alg_params = self.cuml_params.copy()
+
+        def _construct_sgnn() -> CumlT:
+
+            from cuml.neighbors import NearestNeighbors as SGNN
+
+            nn_object = SGNN(output_type="cupy", **cuml_alg_params)
+
+            return nn_object
+
+        row_number_col = alias.row_number
+        input_col, input_cols = self._get_input_columns()
+        assert input_col is not None or input_cols is not None
+        id_col_name = self.getIdCol()
+
+        bcast_qids = self.bcast_qids
+        bcast_qfeatures = self.bcast_qfeatures
+
+        assert bcast_qids is not None and bcast_qfeatures is not None
+
+        def _transform_internal(
+            nn_object: CumlT, df: Union[pd.DataFrame, np.ndarray]
+        ) -> pd.DataFrame:
+
+            item_row_number = df[row_number_col].to_numpy()
+            item = df.drop(row_number_col, axis=1)  # type: ignore
+            if input_col is not None:
+                assert len(item.columns) == 1
+                item = np.array(list(item[item.columns[0]]), order="C")
+
+            if len(item) == 0:
+                return pd.DataFrame(
+                    {
+                        "query_{id_col_name}": [],
+                        "indices": [],
+                        "distances": [],
+                    }
+                )
+
+            nn_object.fit(item)
+            import cupy as cp
+
+            distances, indices = nn_object.kneighbors(bcast_qfeatures.value)
+
+            if cuml_alg_params["algorithm"] == "ivfflat":
+                distances = distances * distances
```

Comment:
Probably good to add a comment for this.   Also, are other distances besides Euclidean supported?

Reply:
Added comments to explain this.